### PR TITLE
RPP - Add interface and concrete classes for Payment Method

### LINKED ISSUE #6681: Payment Process: Payment Method Interface
### Description

Create an interface for payment methods within the payment process. It will be later implemented by the new and saved payment methods.

### Acceptance criteria

The interface should define the following methods:

- `get_id()`, which returns a string identifier of the payment method.
- `get_data()` which returns a serializable array.

`set_data()` should not be needed, as details will be provided through [a factory](https://github.com/Automattic/woocommerce-payments/issues/6685) to the constructor in a different issue.

### Testing instructions

It's an interface, no tests needed.

### Dev notes

Something similar was already implemented in https://github.com/Automattic/woocommerce-payments/pull/6287/files#diff-831ad4ca30978381104cb04449e2c6a1043aecb9ce369323595c06a5d89074fb 

### Additional context

Epic: https://github.com/Automattic/woocommerce-payments/issues/6673
Project: paJDYF-9hL-p2

### LINKED ISSUE #6682: Payment Process: New Payment Method Class
### Description

Extend the [payment method interface](https://github.com/Automattic/woocommerce-payments/issues/6681) for non-saved payment methods.

### Acceptance criteria

- There is a new class, which contains the ID of an unsaved payment method.

### Testing instructions

Unit tests should cover:

- Instantiating with an ID, and retrieving it through `get_id()`.
- Making sure that the data, returned by `get_data()` is correct.

### Dev notes

Previously implemented in https://github.com/Automattic/woocommerce-payments/pull/6287/files#diff-0e50df1381b6ba77df3ed787e6d5aba3dd64bb6a40c601eaa9c0a968fd14a500

### Additional context

Epic: https://github.com/Automattic/woocommerce-payments/issues/6673
Project: paJDYF-9hL-p2

### LINKED ISSUE #6683: Payment Process: Saved Payment Method Class
### Description

Extend the [payment method interface](https://github.com/Automattic/woocommerce-payments/issues/6681) for saved payment methods.

### Acceptance criteria

- There is a new class, which is instantiated with a `WC_Payment_Token` object, and stores the token internally.
- `get_id()` returns the token object's token.
- `get_data()` exports the ID of the token.

### Testing instructions

Unit tests should cover:

- Instantiating with a token, and retrieving it the payment method through `get_id()`.
- Making sure that the data, returned by `get_data()` is correct.

### Dev notes

Previously implemented in https://github.com/Automattic/woocommerce-payments/pull/6287/files#diff-b87a87f2f4e4f25bc0060bd1700fad6b0e007ed2091e6469a89a96001a16a7da

### Additional context

Epic: https://github.com/Automattic/woocommerce-payments/issues/6673
Project: paJDYF-9hL-p2

### PR DESCRIPTION
**Pull request by htdat**
Note: RPP = Re-engineering Payment Process.

Fixes #6681
Fixes #6682 
Fixes #6683

## ❓ Question

Although this PR uses the `Payment Method` term, but I'd suggest using `token` such as `new token` and `reusable token` as mentioned in paJDYF-8J2-p2, especially these parts of that P2 3181a-pb/#plain

#### Changes proposed in this Pull Request

- Add PaymentMethodInterface, and two concrete classes implementing it.

#### Testing instructions

* All PHPUnit tests passed. 
* Achieve 100% cover for new files: 

Markup on 2023-08-23 at 11:17:51

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] NO NEED - Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7020` or branch name `rpp/6681-add-payment-method-interface` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=rpp/6681-add-payment-method-interface) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: b9d4a92463953a5e494e9800c0642520248363ef
- Build time: 2023-08-25 03:02:05 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by htdat** [comment]
Ready for review. This PR looks simple but I'd like to get feedback for the `❓ Question` above?

**Comment by htdat** [comment]
> Happy to take this conversation to Slack or P2 if needed.

@marcinbot - let's do it here p1692844497388649-slack-C05997J5PCG

File: src/Internal/Payment/PaymentMethod/NewPaymentMethod.php
Comment (RadoslavGeorgiev) [inline src/Internal/Payment/PaymentMethod/NewPaymentMethod.php]:
```suggestion
	 * External ID of the payment method.
```

I like the description of the constructor parameter better, let's use it here as well.

File: src/Internal/Payment/PaymentMethod/SavedPaymentMethod.php
Comment (RadoslavGeorgiev) [inline src/Internal/Payment/PaymentMethod/SavedPaymentMethod.php]: I don't think `$this->get_id()` is suitable here. To construct the object from this data, we'd need the token ID.
```suggestion
			'id'   => $this->token->get_id(),
```

**Review by RadoslavGeorgiev — APPROVED** [review]
Great job, the changes look good to me. 🎉 

I only have a couple of small suggestions, feel free to consider them and merge without another review round.

The conversation regarding naming will take a while, and we can always change them here. Let's merge without waiting for a decision.

File: src/Internal/Payment/PaymentMethod/NewPaymentMethod.php
Comment (htdat) [reply src/Internal/Payment/PaymentMethod/NewPaymentMethod.php]: Addressed https://github.com/Automattic/woocommerce-payments/pull/7020/commits/080ed7b9a384a64332319cdcd7a4d716d66d5871

File: src/Internal/Payment/PaymentMethod/SavedPaymentMethod.php
Comment (htdat) [reply src/Internal/Payment/PaymentMethod/SavedPaymentMethod.php]: Good point. Addressed https://github.com/Automattic/woocommerce-payments/pull/7020/commits/b9d4a92463953a5e494e9800c0642520248363ef

**Pull request merged**
